### PR TITLE
Check that all hosts in vC cluster are in the vDS

### LINKED ISSUE #1648: Failed deployment due to hosts not attached to DVS leaves vApp hanging around
**VIC Version:**

0.1 / master / 3190

When deploying a VIC to an VC but not all the hosts are connected to the DVS, the vApp is created but is not cleaned up. This means that future deployments won't work with the same name because  the vApp exists.

empty vapp

```
root@photon-NaTv5i8IA [ /workspace/vic ]# ./vic-machine-linux create --bridge-network Bridge-DPG --image-datastore vsanDatastore  -t 'administrator@vsphere.local:VMware123!@10.27.51.103' --compute-resource Mgmt
INFO[2016-07-25T09:37:17Z] ### Installing VCH ####
INFO[2016-07-25T09:37:17Z] Generating certificate/key pair - private key in ./virtual-container-host-key.pem
INFO[2016-07-25T09:37:18Z] Validating supplied configuration
INFO[2016-07-25T09:37:18Z] Firewall status: ENABLED on "/CNA-DC/host/Mgmt/10.27.51.10"
INFO[2016-07-25T09:37:18Z] Firewall status: ENABLED on "/CNA-DC/host/Mgmt/10.27.51.8"
INFO[2016-07-25T09:37:18Z] Firewall status: ENABLED on "/CNA-DC/host/Mgmt/10.27.51.9"
INFO[2016-07-25T09:37:18Z] Firewall configuration OK on hosts:
INFO[2016-07-25T09:37:18Z]   "/CNA-DC/host/Mgmt/10.27.51.10"
INFO[2016-07-25T09:37:18Z]   "/CNA-DC/host/Mgmt/10.27.51.8"
INFO[2016-07-25T09:37:18Z]   "/CNA-DC/host/Mgmt/10.27.51.9"
INFO[2016-07-25T09:37:18Z] License check OK on hosts:
INFO[2016-07-25T09:37:18Z]   "/CNA-DC/host/Mgmt/10.27.51.10"
INFO[2016-07-25T09:37:18Z]   "/CNA-DC/host/Mgmt/10.27.51.8"
INFO[2016-07-25T09:37:18Z]   "/CNA-DC/host/Mgmt/10.27.51.9"
INFO[2016-07-25T09:37:18Z] DRS check OK on:
INFO[2016-07-25T09:37:18Z]   "/CNA-DC/host/Mgmt/Resources"
INFO[2016-07-25T09:37:18Z] Creating virtual app "virtual-container-host"
INFO[2016-07-25T09:37:18Z] Creating appliance on target
INFO[2016-07-25T09:37:18Z] Network role "external" is sharing NIC with "management"
INFO[2016-07-25T09:37:18Z] Network role "client" is sharing NIC with "management"
ERRO[2016-07-25T09:37:20Z] Operation failed: A specified parameter was not correct: spec.deviceChange.device.port.switchUuid - (&{{} %!!(MISSING)s(*types.InvalidArgument=&{{{<nil> [{{} com.vmware.vim.vpxd.dvs.nonHostMember.label [{{} host 10.27.51.8} {{} dvs DSwitch}] Host 10.27.51.8 is not a member of VDS DSwitch}]}} spec.deviceChange.device.port.switchUuid}) A specified parameter was not correct: spec.deviceChange.device.port.switchUuid})
ERRO[2016-07-25T09:37:20Z] Unable to create appliance VM: Operation failed: A specified parameter was not correct: spec.deviceChange.device.port.switchUuid - (&{{} %!s(*types.InvalidArgument=&{{{<nil> [{{} com.vmware.vim.vpxd.dvs.nonHostMember.label [{{} host 10.27.51.8} {{} dvs DSwitch}] Host 10.27.51.8 is not a member of VDS DSwitch}]}} spec.deviceChange.device.port.switchUuid}) A specified parameter was not correct: spec.deviceChange.device.port.switchUuid})
INFO[2016-07-25T09:37:20Z] Collecting a0522635-224e-4171-92b1-7edaf726b39f vpxd.log
ERRO[2016-07-25T09:37:21Z] --------------------
ERRO[2016-07-25T09:37:21Z] vic-machine-linux failed: Creating the appliance failed with Operation failed: A specified parameter was not correct: spec.deviceChange.device.port.switchUuid - (&{{} %!s(*types.InvalidArgument=&{{{<nil> [{{} com.vmware.vim.vpxd.dvs.nonHostMember.label [{{} host 10.27.51.8} {{} dvs DSwitch}] Host 10.27.51.8 is not a member of VDS DSwitch}]}} spec.deviceChange.device.port.switchUuid}) A specified parameter was not correct: spec.deviceChange.device.port.switchUuid}). Exiting...

root@photon-NaTv5i8IA [ /workspace/vic ]# ./vic-machine-linux create --bridge-network Bridge-DPG --image-datastore vsanDatastore  -t 'administrator@vsphere.local:VMware123!@10.27.51.103' --compute-resource Mgmt
INFO[2016-07-25T09:51:40Z] ### Installing VCH ####
INFO[2016-07-25T09:51:40Z] Generating certificate/key pair - private key in ./virtual-container-host-key.pem
INFO[2016-07-25T09:51:41Z] Validating supplied configuration
INFO[2016-07-25T09:51:41Z] Firewall status: ENABLED on "/CNA-DC/host/Mgmt/10.27.51.10"
INFO[2016-07-25T09:51:41Z] Firewall status: ENABLED on "/CNA-DC/host/Mgmt/10.27.51.8"
INFO[2016-07-25T09:51:41Z] Firewall status: ENABLED on "/CNA-DC/host/Mgmt/10.27.51.9"
INFO[2016-07-25T09:51:41Z] Firewall configuration OK on hosts:
INFO[2016-07-25T09:51:41Z]   "/CNA-DC/host/Mgmt/10.27.51.10"
INFO[2016-07-25T09:51:41Z]   "/CNA-DC/host/Mgmt/10.27.51.8"
INFO[2016-07-25T09:51:41Z]   "/CNA-DC/host/Mgmt/10.27.51.9"
INFO[2016-07-25T09:51:42Z] License check OK on hosts:
INFO[2016-07-25T09:51:42Z]   "/CNA-DC/host/Mgmt/10.27.51.10"
INFO[2016-07-25T09:51:42Z]   "/CNA-DC/host/Mgmt/10.27.51.8"
INFO[2016-07-25T09:51:42Z]   "/CNA-DC/host/Mgmt/10.27.51.9"
INFO[2016-07-25T09:51:42Z] DRS check OK on:
INFO[2016-07-25T09:51:42Z]   "/CNA-DC/host/Mgmt/Resources"
ERRO[2016-07-25T09:51:43Z] virtual app "/CNA-DC/host/Mgmt/Resources/virtual-container-host" is found, but is not VCH, please choose different name
INFO[2016-07-25T09:51:43Z] Collecting a0522635-224e-4171-92b1-7edaf726b39f vpxd.log
ERRO[2016-07-25T09:51:43Z] --------------------
ERRO[2016-07-25T09:51:43Z] vic-machine-linux failed: virtual app "/CNA-DC/host/Mgmt/Resources/virtual-container-host" is found, but is not VCH, please choose different name

root@photon-NaTv5i8IA [ /workspace/vic ]#
```

**Expected behavior:**
On failure to deploy, remove previously created vApp

**Actual behavior:**
As above

### PR DESCRIPTION
**Pull request by andrewtchin**
Prevent VCH install on vCenter if the specified network is not connected to all hosts in the cluster

Fixes #1648, #802 

```
⇒  bin/vic-machine-linux create --target 10.17.109.115 --image-datastore vsanDatastore --name chin-vic-15 --user administrator@vsphere.local --password password --compute-resource cluster1/chin --bridge-network vch-test
INFO[2016-07-29T08:43:51-07:00] ### Installing VCH ####                      
INFO[2016-07-29T08:43:51-07:00] Generating certificate/key pair - private key in ./chin-vic-15-key.pem 
INFO[2016-07-29T08:43:52-07:00] Validating supplied configuration            
INFO[2016-07-29T08:43:54-07:00] vDS configuration OK on "vch-test"           
INFO[2016-07-29T08:43:54-07:00] Firewall status: ENABLED on "/dc1/host/cluster1/10.17.109.74" 
INFO[2016-07-29T08:43:54-07:00] Firewall status: ENABLED on "/dc1/host/cluster1/10.17.109.76" 
INFO[2016-07-29T08:43:55-07:00] Firewall status: ENABLED on "/dc1/host/cluster1/10.17.109.169" 
INFO[2016-07-29T08:43:55-07:00] Firewall configuration OK on hosts:          
INFO[2016-07-29T08:43:55-07:00]   "/dc1/host/cluster1/10.17.109.74"          
INFO[2016-07-29T08:43:55-07:00]   "/dc1/host/cluster1/10.17.109.76"          
INFO[2016-07-29T08:43:55-07:00]   "/dc1/host/cluster1/10.17.109.169"         
INFO[2016-07-29T08:43:55-07:00] License check OK on hosts:                   
INFO[2016-07-29T08:43:55-07:00]   "/dc1/host/cluster1/10.17.109.74"          
INFO[2016-07-29T08:43:55-07:00]   "/dc1/host/cluster1/10.17.109.76"          
INFO[2016-07-29T08:43:55-07:00]   "/dc1/host/cluster1/10.17.109.169"         
INFO[2016-07-29T08:43:55-07:00] DRS check OK on:                             
INFO[2016-07-29T08:43:55-07:00]   "/dc1/host/cluster1/Resources/chin"        
INFO[2016-07-29T08:43:57-07:00] Creating virtual app "chin-vic-15"           
INFO[2016-07-29T08:43:57-07:00] Creating appliance on target                 
INFO[2016-07-29T08:43:57-07:00] Network role "client" is sharing NIC with "external" 
INFO[2016-07-29T08:43:57-07:00] Network role "management" is sharing NIC with "external" 
INFO[2016-07-29T08:44:12-07:00] Uploading images for container               
INFO[2016-07-29T08:44:12-07:00] 	"/home/chin/go/src/github.com/vmware/vic/bin/appliance.iso" 
INFO[2016-07-29T08:44:12-07:00] 	"/home/chin/go/src/github.com/vmware/vic/bin/bootstrap.iso" 
INFO[2016-07-29T08:44:52-07:00] Registering VCH as a vSphere extension       
INFO[2016-07-29T08:45:04-07:00] Waiting for IP information                   
INFO[2016-07-29T08:45:38-07:00] Waiting for major appliance components to launch 
INFO[2016-07-29T08:45:48-07:00] Initialization of appliance successful       
INFO[2016-07-29T08:45:48-07:00]                                              
INFO[2016-07-29T08:45:48-07:00] Log server:                                  
INFO[2016-07-29T08:45:48-07:00] https://10.17.109.112:2378                   
INFO[2016-07-29T08:45:48-07:00]                                              
INFO[2016-07-29T08:45:48-07:00] DOCKER_HOST=10.17.109.112:2376               
INFO[2016-07-29T08:45:48-07:00]                                              
INFO[2016-07-29T08:45:48-07:00] Connect to docker:                           
INFO[2016-07-29T08:45:48-07:00] docker -H 10.17.109.112:2376 --tls info      
INFO[2016-07-29T08:45:48-07:00] Installer completed successfully 
```

```
⇒  bin/vic-machine-linux create --target 10.17.109.115 --image-datastore vsanDatastore --name chin-vic-15 --user administrator@vsphere.local --password password --compute-resource cluster1/chin --bridge-network vch-test
INFO[2016-07-29T08:49:21-07:00] ### Installing VCH ####                      
INFO[2016-07-29T08:49:21-07:00] Generating certificate/key pair - private key in ./chin-vic-15-key.pem 
INFO[2016-07-29T08:49:22-07:00] Validating supplied configuration            
ERRO[2016-07-29T08:49:24-07:00] vDS configuration incorrect on "vch-test". All cluster hosts must be in the vDS. 
ERRO[2016-07-29T08:49:24-07:00]   "vch-test" is missing hosts:               
ERRO[2016-07-29T08:49:24-07:00]     "/dc1/host/cluster1/10.17.109.169"       
INFO[2016-07-29T08:49:24-07:00] Firewall status: ENABLED on "/dc1/host/cluster1/10.17.109.74" 
INFO[2016-07-29T08:49:24-07:00] Firewall status: ENABLED on "/dc1/host/cluster1/10.17.109.76" 
INFO[2016-07-29T08:49:25-07:00] Firewall status: ENABLED on "/dc1/host/cluster1/10.17.109.169" 
INFO[2016-07-29T08:49:25-07:00] Firewall configuration OK on hosts:          
INFO[2016-07-29T08:49:25-07:00]   "/dc1/host/cluster1/10.17.109.74"          
INFO[2016-07-29T08:49:25-07:00]   "/dc1/host/cluster1/10.17.109.76"          
INFO[2016-07-29T08:49:25-07:00]   "/dc1/host/cluster1/10.17.109.169"         
INFO[2016-07-29T08:49:25-07:00] License check OK on hosts:                   
INFO[2016-07-29T08:49:25-07:00]   "/dc1/host/cluster1/10.17.109.74"          
INFO[2016-07-29T08:49:25-07:00]   "/dc1/host/cluster1/10.17.109.76"          
INFO[2016-07-29T08:49:25-07:00]   "/dc1/host/cluster1/10.17.109.169"         
INFO[2016-07-29T08:49:25-07:00] DRS check OK on:                             
INFO[2016-07-29T08:49:25-07:00]   "/dc1/host/cluster1/Resources/chin"        
ERRO[2016-07-29T08:49:25-07:00] --------------------                         
ERRO[2016-07-29T08:49:25-07:00] All cluster hosts must be in the vDS. "vch-test" is missing hosts: [/dc1/host/cluster1/10.17.109.169] 
ERRO[2016-07-29T08:49:25-07:00] Create cannot continue: configuration validation failed 
ERRO[2016-07-29T08:49:25-07:00] --------------------                         
ERRO[2016-07-29T08:49:25-07:00] vic-machine-linux failed: validation of configuration failed
```

### CODE REVIEW

File: lib/install/validate/validator.go [hunk 1]
```diff
@@ -267,6 +267,65 @@ func (v *Validator) storage(ctx context.Context, input *data.Data, conf *config.
...
+
+	nets, err := v.getNetworks(ctx, network)
+	if err != nil {
+		return err
+	}
+	if len(nets) > 1 {
+		// TODO: error about required disabmiguation and list entries in nets
+		return errors.New("ambiguous network " + network)
+	}
+	ref := nets[0].Reference()
```
Comment (jzt) [inline lib/install/validate/validator.go]: Maybe a dumb question, but could `len(nets)` be 0?

File: lib/install/validate/validator.go [hunk 2]
```diff
@@ -267,6 +267,65 @@ func (v *Validator) storage(ctx context.Context, input *data.Data, conf *config.
...
+	if !v.IsVC() {
+		return nil
+	}
+
+	clusterHosts, err := v.Session.Cluster.Hosts(ctx)
+	if err != nil {
+		return err
+	}
+
+	nets, err := v.getNetworks(ctx, network)
```
Comment (emlin) [inline lib/install/validate/validator.go]: you can use dpgHelper here

**Comment by jzt** [comment]
LGTM 🚢 🎱

File: lib/install/validate/validator.go [hunk 3]
```diff
@@ -368,6 +427,13 @@ func (v *Validator) network(ctx context.Context, input *data.Data, conf *config.
 	// port forwarding
 	conf.AddNetwork(bridgeNet)
 
+	if v.IsVC() {
+		err := v.checkVDSMembership(ctx, input.BridgeNetworkName)
```
Comment (emlin) [inline lib/install/validate/validator.go]: Cause your method is always checked after the network validation is done, so suppose you could pass in the dpg reference directly into your function.

File: lib/install/validate/validator.go [hunk 4]
```diff
@@ -430,6 +496,12 @@ func (v *Validator) network(ctx context.Context, input *data.Data, conf *config.
 		if input.BridgeNetworkName == net {
 			v.NoteIssue(errors.Errorf("the bridge network must not be shared with another network role - %q also mapped as container network %q", input.BridgeNetworkName, name))
 		}
+		if v.IsVC() {
```
Comment (emlin) [inline lib/install/validate/validator.go]: have this detection inside of your method, could make the caller clearer

File: tests/test-cases/Group6-VIC-Machine/6-5-Create-Validation.md [hunk 5]
```diff
@@ -10,32 +10,66 @@ Verify vic-machine create validation functions, this does not include validation
...
+2. Issue the following command:
+```
+vic-machine create --name=<VCH_NAME> --target=<TEST_URL> \
+    --user=<TEST_USERNAME> --image-datastore=<TEST_DATASTORE> --password=<TEST_PASSWORD> \
+    --bridge-network=<NETWORK> --compute-resource=<TEST_RESOURCE>
+```
+
+#Expected Outcome:
+* Output contains message indicating vDS configuration OK
+* Deployment succeeds
```
Comment (emlin) [inline tests/test-cases/Group6-VIC-Machine/6-5-Create-Validation.md]: need to run regression test here

File: lib/install/validate/validator.go (hunk 1 above)
Comment (andrewtchin) [inline lib/install/validate/validator.go]: no getNetwork has err if network isn't found

File: lib/install/validate/validator.go [hunk 6]
```diff
@@ -290,15 +290,11 @@ func (v *Validator) checkVDSMembership(ctx context.Context, network string) erro
 		return err
 	}
 
-	nets, err := v.getNetworks(ctx, network)
+	net, err := v.getNetwork(ctx, network)
```
Comment (emlin) [inline lib/install/validate/validator.go]: alright, I think you don't have to call this method in this membership checking, cause all networks are already checked ahead of your method call

File: lib/install/validate/validator.go [hunk 7]
```diff
@@ -427,11 +423,9 @@ func (v *Validator) network(ctx context.Context, input *data.Data, conf *config.
 	// port forwarding
 	conf.AddNetwork(bridgeNet)
 
-	if v.IsVC() {
-		err := v.checkVDSMembership(ctx, input.BridgeNetworkName)
-		if err != nil {
-			v.NoteIssue(fmt.Errorf("Unable to check hosts in vDS for %q: %s", input.BridgeNetworkName, err))
-		}
+	err = v.checkVDSMembership(ctx, input.BridgeNetworkName)
```
Comment (emlin) [inline lib/install/validate/validator.go]: https://github.com/vmware/vic/pull/1747/commits/7dd9118932c06e86921cf27470115c6414cf1a2e#diff-a7c93db9877e86593965b23165a2b99fR394
is where the network is checked, so I think you can just use endpointMoref

File: lib/install/validate/validator.go [hunk 8]
```diff
@@ -267,6 +267,55 @@ func (v *Validator) storage(ctx context.Context, input *data.Data, conf *config.
 	}
 }
 
+func (v *Validator) inDVP(host types.ManagedObjectReference, dvpHosts []types.ManagedObjectReference) bool {
```
Comment (hmahmood) [inline lib/install/validate/validator.go]: A comment here would be helpful; not readily apparent what a DVP is.

**Comment by hmahmood** [comment]
lgtm

**Comment by emlin** [comment]
lgtm 👍

**Comment by andrewtchin** [comment]
I'm having trouble testing this on vC while remote, waiting until I'm on the LAN to test before merging

**Comment by andrewtchin** [comment]
♻️

**Pull request merged**